### PR TITLE
feat: Typeahead searchable filter_box for dashboard

### PR DESCRIPTION
**Pull request by graceguo-supercat**
### SUMMARY
Currently Superset only show top 1000 options for a dropdown list in filter_box component. If a filter has > 1000 options, user won't see all the options. When user type-in with correct value (space, capital letter etc), the filter value can still get applied. But this experience confused many users. When they didn't see the option from the list, they thought it was not an option.

This PR is to enable async searchable options for filter_box in Superset dashboard:
- by default, filter_box only fetch top 1000 options (like right now).
- if enable `Search All Filter Options` and this filter key has >= 1000 options, filter box will load options from a remote source as the user types, with 1000 options per load.
- when user typing in, filter_box is actually sending out a new query with extra sql expression ` WHERE lower(_column_key_) like '_input_value_'`

Screen Shot 2020-07-07 at 9 29 35 AM

For example: Malibu is not in the top 1000 destination.
**Before: user cannot see Malibu is an option** 

42cSlzWfZb-2

**After: Malibu will show up while user typing**
PIvGCagMOq

### TEST PLAN
Manual test

### ADDITIONAL INFORMATION

- [x] Changes UI

### CODE REVIEW

**Comment by mistercrunch** [comment]
The name `Fixed Filter Options` isn't very clear. I'm not sure what to call it though.

File: superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx [hunk 1]
```diff
@@ -202,6 +220,21 @@ export default class FilterBoxItemControl extends React.Component {
             />
           }
         />
+        <FormRow
+          label={t('Fixed Filter Options')}
+          tooltip={t(
+            'By default, filter box will only load top 1000 filter options once. ' +
+              'When disabled, filter box will load options from a remote source as the user types, ' +
+              'with 1000 options per load.',
```
Comment (ktmud) [inline superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx:228]: Should we make the async loading the default or allow Superset admins to change it?

Comment (graceguo-supercat) [reply superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx:228]: let me check with PM tomorrow.

File: superset-frontend/src/visualizations/FilterBox/FilterBox.jsx [hunk 2]
```diff
@@ -162,6 +178,75 @@ class FilterBox extends React.Component {
...
   }
 
+  /**
+   * Generate a debounce function that loads options for a specific column
+   */
+  debounceLoadOptions(key) {
+    if (!(key in this.debouncerCache)) {
+      this.debouncerCache[key] = debounce((input, callback) => {
+        this.loadOptions(key, input).then(callback);
+      }, 500);
```
Comment (graceguo-supercat) [inline superset-frontend/src/visualizations/FilterBox/FilterBox.jsx:188]: use a large delay to reduce number of queries.

**Comment by graceguo-supercat** [comment]
> The name `Fixed Filter Options` isn't very clear. How about `Dynamic Autocomplete` ?

even with fixed number of options, when user type in, filter_box can do `autocomplete`. But it only auto-completes with options already returned to front-end.

**Comment by rusackas** [comment]
> The name `Fixed Filter Options` isn't very clear. How about `Dynamic Autocomplete` ?

Names aside, I like the idea of checking the box to turn on the query feature, rather than turning on the limitation.

From a UX perspective, having that enabled by default seems like it would cause less pain/confusion, if it's not too "expensive"

And as for the name, that's tough... maybe "Search All Filter Options?"

**Comment by graceguo-supercat** [comment]
@mistercrunch @rusackas @zuzana-vej @sylvia-tomiyama I would like to hear your suggestion about:

- the control name `Fixed Filter Options` and description.
- when this feature should be enabled? Currently it is an option and chart owner can set it for a filter key. @ktmud suggest we should auto enable this when filter has > 1000 options. My concern is because auto-completes trigger a new query, if the datasource is made of many table joins, the response will be very slow. If we automatically enable this for all large filter box, there is no way for slow ones being escaped.
- Note: auto-completes only happens when user type-in. If user only load page, open filter_box, play with options in the initial load, this will not trigger extra query.

**Comment by ktmud** [comment]
> The name `Fixed Filter Options` isn't very clear. How about `Dynamic Autocomplete` ?

+1 to "Dynamic Autocomplete".

File: superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx (hunk 1 above)
Comment (mistercrunch) [reply superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx:228]: !danger here, it's very possible that you'd be running dozens of `SELECT DISTINCT dimension_column FROM VERY_LARGE_TABLE WHERE dimension_column LIKE '%{str}%'`.  Your poor Presto cluster might suffer.

Also the cache hit rate would be very low compared to the normal filter.

**Comment by mistercrunch** [comment]
I want to make sure people see this comment https://github.com/apache/incubator-superset/pull/10210#discussion_r449391782

File: superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx (hunk 1 above)
Comment (mistercrunch) [reply superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx:228]: So I'd argue for unchecked by default, and a good explanation of the tradeoff in a toolitp

**Comment by graceguo-supercat** [comment]
we like `Search All Filter Options`. also updated description:
Screen Shot 2020-07-06 at 10 27 29 AM

As description said, uncheck by default, `and` only send extra query when a column has > 1000 initial options.

File: superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx [hunk 3]
```diff
@@ -202,6 +220,21 @@ export default class FilterBoxItemControl extends React.Component {
             />
           }
         />
+        <FormRow
+          label={t('Search All Filter Options')}
+          tooltip={t(
+            'By default, filter box loads top 1000 records. Check this box if you want to ' +
+              'enable dynamically searching and loading filter values as user types ' +
+              '(applicable only when the filter returns more than 1000 records).',
```
Comment (ktmud) [inline superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx:228]: Suggestion:

> By default, each filter loads only at most 1000 choices at the initial page load. Check this box if you have more than 1000 filter values and want to enable dynamically searching that loads filter values as users type.

**Review by ktmud — APPROVED** [review]

File: superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx (hunk 3 above)
Comment (graceguo-supercat) [reply superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx]: description is updated.

**Pull request merged**